### PR TITLE
Validate emails

### LINKED ISSUE #17: Validate email format upon registering candidates and voters
Currently, voters emails (which are used to send voting tokens) are not validated.
As a consequence, when we send voting tokens, we may crash (as the postal jar we use cannot handle that).
Realistically, we shouldn't allow for candidates or voters to have emails that are not even in a valid email format.

### PR DESCRIPTION
**Pull request by Kineolyan**
I added some validation when registering voters. This PR checks that the list of voters does not contain any voter with an invalid email.
Otherwise, the list of invalid voters is returned and the process of registration is aborted.

As there were no tests on the controller, I haven't check the operation through the request. But there are unit tests on the regexp.

Closes: #17.

### CODE REVIEW

File: src/clj/election/controllers/elections.clj [hunk 1]
```diff
@@ -59,15 +59,35 @@
...
   (let [election (db/election (Integer. election-id))
     response (response/redirect (paths/election-path election-id))]
     (if (auth/can-register-voters? election user)
       (let [voters (voters-from voters-file)
-        added-voters-count (voters/register-voters (:id election) voters)]
-        (if added-voters-count
-          (assoc response :flash {:type :notice :message (i18n/t request :elections/new-voters-registered added-voters-count)})
-          (assoc response :flash {:type :error :message (i18n/t request :elections/voter-registration-failed)})
+            invalid-voters (get-invalid-mails voters)]
+        (if (empty? invalid-voters)
```
Comment (hugocorbucci) [inline src/clj/election/controllers/elections.clj]: Any chance we could make this report on the invalid ones but still add the valid ones?
I think it wouldn't be too hard

File: test/clj/election/controllers/elections_test.clj [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    (let [result (e/is-valid-email? "someone@gmail.com")]
+      (is result)
+    )
+  )
+  (testing "accepts email with dots"
+    (let [result (e/is-valid-email? "a.b.c@yahoo.com")]
+      (is result)
+    )
+  )
+  (testing "accepts special gmail with +"
```
Comment (hugocorbucci) [inline test/clj/election/controllers/elections_test.clj:17]: Mind adding a couple more tests for other than `.com` ending domains?
Like `agilealliancebrazil.org.br`

**Review by hugocorbucci — CHANGES_REQUESTED** [review]
Thank you very much for this! This looks great!
I had a couple requests before I merge this but this is already fantastic!

File: test/clj/election/controllers/elections_test.clj (hunk 2 above)
Comment (Kineolyan) [reply test/clj/election/controllers/elections_test.clj:17]: Sure. I will do so.

File: src/clj/election/controllers/elections.clj (hunk 1 above)
Comment (Kineolyan) [reply src/clj/election/controllers/elections.clj]: Yes I can. My only concern that then, you would have to create a new file for the invalid voters spotted, as all other voters would have already been registered.
Currently, it just works or you have to retry with a valid list.
But I guess that it is pretty easy to create a new file with the corrected emails, so I will do as you suggest.

Comment (hugocorbucci) [reply src/clj/election/controllers/elections.clj]: The voter addition is actually idempotent so already registered voters are ignored. So you just need to fix the emails that were wrong.

Comment (Kineolyan) [reply src/clj/election/controllers/elections.clj]: Perfect. I will do so.

**Comment by Kineolyan** [comment]
Now, I continue the registration of the valid emails.
I also added more tests on valid emails, following some lists of valid/invalid emails from the Internet.
I documented in the tests the cases causing issues to the regexp. From what I see, the errors are not caused by standard emails, so we should be good.

**Comment by Kineolyan** [comment]
Also, note that I added an entry in the pt_BR translations, but as I don't speak Brazilian, it is still in english. You may want to update it or provide me with the correct translation, before we merge.

**Comment by hugocorbucci** [comment]
```
:some-voters-registered "%d novo(s) eleitor(es) inscrito(s) (existentes: %d, inválidos: %d)"
```
Thank you very much @Kineolyan! This is amazing!

**Pull request merged**